src/vos/vos_dtx.c

```diff
@@ -976,84 +1482,97 @@ vos_dtx_abort(daos_handle_t coh, daos_epoch_t epoch, struct dtx_id *dtis,
 
 	/* Abort multiple DTXs via single PMDK transaction. */
 	rc = vos_tx_begin(vos_cont2umm(cont));
-	if (rc != 0)
-		return rc;
+	if (rc == 0) {
+		for (i = 0; i < count; i++)
+			vos_dtx_abort_one(cont, epoch, &dtis[i]);
 
-	for (i = 0; rc == 0 && i < count; i++)
-		rc = vos_dtx_abort_one(cont, epoch, &dtis[i], force);
+		rc = vos_tx_end(vos_cont2umm(cont), 0);
+	}
 
-	return vos_tx_end(vos_cont2umm(cont), rc);
+	return rc;
 }
 
 int
-vos_dtx_aggregate(daos_handle_t coh, uint64_t max, uint64_t age)
+vos_dtx_aggregate(daos_handle_t coh)
 {
 	struct vos_container		*cont;
+	struct vos_cont_df		*cont_df;
 	struct umem_instance		*umm;
-	struct vos_dtx_table_df		*tab;
-	umem_off_t			 dtx_umoff;
-	uint64_t			 count;
-	int				 rc = 0;
+	struct vos_dtx_scm_blob		*dsb;
+	struct vos_dtx_scm_blob		*tmp;
+	umem_off_t			 dsb_off;
+	int				 rc;
+	int				 i;
 
 	cont = vos_hdl2cont(coh);
 	D_ASSERT(cont != NULL);
 
-	umm = &cont->vc_pool->vp_umm;
-	tab = &cont->vc_cont_df->cd_dtx_table_df;
+	umm = vos_cont2umm(cont);
+	cont_df = cont->vc_cont_df;
 
-	rc = vos_tx_begin(vos_cont2umm(cont));
+	dsb_off = cont_df->cd_dtx_committed_head;
+	dsb = umem_off2ptr(umm, dsb_off);
+	if (dsb == NULL || dsb->dsb_count == 0)
+		return 0;
+
+	rc = vos_tx_begin(umm);
 	if (rc != 0)
 		return rc;
 
-	umem_tx_add_ptr(umm, tab, sizeof(*tab));
-	for (count = 0, dtx_umoff = tab->tt_entry_head;
-	     count < max && !dtx_is_null(dtx_umoff); count++) {
-		struct vos_dtx_entry_df	*dtx;
-		d_iov_t		 kiov;
-		umem_off_t		 umoff;
+	for (i = 0; i < dsb->dsb_count &&
+	     !d_list_empty(&cont->vc_dtx_committed_list); i++) {
```

Comment:
dbs_count or committed_list is always is relative small so that the original 'max' isn't necessary?

Reply:
Each time the DTX aggregation will handle one vos_dtx_scm_blob that contains thousands of committed DTX entry, then the whole vos_dtx_scm_blob will be released. That will avoid the on-disk changes for per committed DTX entry.